[gradle] Concurrency issue with many simultaneous builds

<!-- Thanks for filing an issue! Please use this template to provide more context for the issue . -->

**Description of the issue**: In a project that can run dozens of jib build tasks concurrently, an exception was thrown when working with the cache

```
Step #3 - "curio-generated-build-all": Caused by: com.google.cloud.tools.jib.plugins.common.BuildStepsExecutionException: Build image failed
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.plugins.common.BuildStepsRunner.build(BuildStepsRunner.java:269)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.gradle.BuildImageTask.buildImage(BuildImageTask.java:119)
Step #3 - "curio-generated-build-all": at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Step #3 - "curio-generated-build-all": at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Step #3 - "curio-generated-build-all": at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Step #3 - "curio-generated-build-all": at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
Step #3 - "curio-generated-build-all": ... 133 more
Step #3 - "curio-generated-build-all": Caused by: java.nio.file.FileSystemException: /root/.cache/google-cloud-tools-java/jib/tmp/10954747174794241924 -> /root/.cache/google-cloud-tools-java/jib/layers/096ebeedeeb6534eadf6d1ae8173145e8f43aedf29f5ae179f8b2e899f84f344: Directory not empty
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.cache.DefaultCacheStorageWriter.moveIfDoesNotExist(DefaultCacheStorageWriter.java:75)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.cache.DefaultCacheStorageWriter.write(DefaultCacheStorageWriter.java:154)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.cache.DefaultCacheStorage.write(DefaultCacheStorage.java:74)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.cache.Cache.writeCompressedLayer(Cache.java:68)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.builder.steps.PullAndCacheBaseImageLayerStep.call(PullAndCacheBaseImageLayerStep.java:81)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.builder.steps.PullAndCacheBaseImageLayerStep.call(PullAndCacheBaseImageLayerStep.java:36)
Step #3 - "curio-generated-build-all": at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
Step #3 - "curio-generated-build-all": at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
Step #3 - "curio-generated-build-all": at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
Step #3 - "curio-generated-build-all": 
```

**Expected behavior**: The build succeeds

**Steps to reproduce**:  This only happens sometimes, and on a build with dozens of projects. I suspect it will be hard to reproduce with a small build.

**Environment**: Linux, Gradle 5.0-rc-4, Java 11, gradle-jib-plugin 0.10.0

**`jib-gradle-plugin` Configuration**: 

Standard configuration, just lots of subprojects

**Log output**:

```
Step #3 - "curio-generated-build-all": Caused by: com.google.cloud.tools.jib.plugins.common.BuildStepsExecutionException: Build image failed
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.plugins.common.BuildStepsRunner.build(BuildStepsRunner.java:269)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.gradle.BuildImageTask.buildImage(BuildImageTask.java:119)
Step #3 - "curio-generated-build-all": at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Step #3 - "curio-generated-build-all": at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Step #3 - "curio-generated-build-all": at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Step #3 - "curio-generated-build-all": at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
Step #3 - "curio-generated-build-all": ... 133 more
Step #3 - "curio-generated-build-all": Caused by: java.nio.file.FileSystemException: /root/.cache/google-cloud-tools-java/jib/tmp/10954747174794241924 -> /root/.cache/google-cloud-tools-java/jib/layers/096ebeedeeb6534eadf6d1ae8173145e8f43aedf29f5ae179f8b2e899f84f344: Directory not empty
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.cache.DefaultCacheStorageWriter.moveIfDoesNotExist(DefaultCacheStorageWriter.java:75)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.cache.DefaultCacheStorageWriter.write(DefaultCacheStorageWriter.java:154)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.cache.DefaultCacheStorage.write(DefaultCacheStorage.java:74)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.cache.Cache.writeCompressedLayer(Cache.java:68)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.builder.steps.PullAndCacheBaseImageLayerStep.call(PullAndCacheBaseImageLayerStep.java:81)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.builder.steps.PullAndCacheBaseImageLayerStep.call(PullAndCacheBaseImageLayerStep.java:36)
Step #3 - "curio-generated-build-all": at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
Step #3 - "curio-generated-build-all": at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
Step #3 - "curio-generated-build-all": at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
Step #3 - "curio-generated-build-all": 
```

**Additional Information**: <!-- Any additional information that may be helpful -->

The OS is Linux (more specifically Google's Cloud Build VM with a build image based on debian), so I think atomic moves should work and would not expect #1044 to be related.

<!-- Thanks for contributing! -->

I managed to coax a build from JitPack but the Maven deems the plugin as invalid as the JitPack coordinates don't match the plugin descriptor's coordinates:
```
[ERROR] Invalid plugin descriptor for com.github.GoogleContainerTools.jib:jib-maven-plugin:i1273-SNAPSHOT (/Users/bsd/.m2/repository/com/github/GoogleContainerTools/jib/jib-maven-plugin/i1273-fdf8a19181-1/jib-maven-plugin-i1273-fdf8a19181-1.jar), Plugin's descriptor contains the wrong group ID: com.google.cloud.tools, Plugin's descriptor contains the wrong version: 0.10.2-SNAPSHOT -> [Help 1]
```
and there [doesn't seem to be any way to ignore this validation](https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/plugin/internal/DefaultMavenPluginManager.java#L250).

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/cache/DefaultCacheStorageWriter.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/cache/DefaultCacheStorageWriter.java
@@ -29,13 +29,10 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.nio.file.AtomicMoveNotSupportedException;
-import java.nio.file.DirectoryNotEmptyException;
-import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.FileSystemException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
 import java.util.zip.GZIPInputStream;
 import java.util.zip.GZIPOutputStream;
 
@@ -67,6 +64,8 @@ class DefaultCacheStorageWriter {
    * @throws IOException if an I/O exception occurs
    */
   private static void moveIfDoesNotExist(Path source, Path destination) throws IOException {
+    // If the file already exists, we skip renaming and use the existing file. This happens if a
+    // new layer happens to have the same content as a previously-cached layer.
     if (Files.exists(destination)) {
       return;
     }
@@ -74,28 +73,9 @@ class DefaultCacheStorageWriter {
     try {
       Files.move(source, destination, StandardCopyOption.ATOMIC_MOVE);
 
-    } catch (FileAlreadyExistsException ignored) {
-      // If the file already exists, we skip renaming and use the existing file. This happens if a
-      // new layer happens to have the same content as a previously-cached layer.
-      //
-      // Do not attempt to remove the try-catch block with the idea of checking file existence
-      // before moving; there can be concurrent file moves.
-
-    } catch (AtomicMoveNotSupportedException ignored) {
-      try {
-        Files.move(source, destination);
-
-      } catch (FileAlreadyExistsException alsoIgnored) {
-        // Same reasoning
-
-      } catch (DirectoryNotEmptyException ex) {
-        // The file system cannot rename directories, so we must resort to copying the directory.
-        Files.createDirectory(destination);
-        try (Stream<Path> sourceFiles = Files.list(source)) {
-          for (Path sourceFile : sourceFiles.collect(Collectors.toList())) {
-            Files.copy(sourceFile, destination.resolve(sourceFile.getFileName()));
-          }
-        }
+    } catch (FileSystemException ex) {
+      if (!Files.exists(destination)) {
+        throw ex;
       }
     }
   }

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/cache/DefaultCacheStorageWriter.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/cache/DefaultCacheStorageWriter.java
@@ -71,7 +71,7 @@ class DefaultCacheStorageWriter {
     }
 
     try {
-      Files.move(source, destination, StandardCopyOption.ATOMIC_MOVE);
+      Files.move(source, destination);
 
     } catch (FileSystemException ex) {
       if (!Files.exists(destination)) {

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/cache/DefaultCacheStorageWriter.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/cache/DefaultCacheStorageWriter.java
@@ -75,6 +75,7 @@ class DefaultCacheStorageWriter {
 
     } catch (FileSystemException ex) {
       if (!Files.exists(destination)) {
+        // TODO to log that the destination exists
         throw ex;
       }
     }

